zip-0221.rst

```diff
@@ -0,0 +1,585 @@
+
+``hashSubtreeCommitment`` forms the strucuture of the commitment tree. Other metadata commitments were chosen to serve specific purposes. Variable-length commitments are placed last, so that most metadata in a node can be directly indexed. We considered using fixed-length commitments here, but opted for variable-length, in order to marginally reduce the memory requirements for managing and updating the commitment trees.
+
+In leaf nodes, some information is repeated. We chose to do this so that leaf nodes could be treated identically to internal and root nodes for all algorithms and (de)serializers. Leaf nodes are easily identifiable, as they will show proof of work in the ``hashSubtreeCommitment`` field, and their block range (calculated as ``nLatestHeight - nEarliestHeight + 1``) will be precisely 1. For the same reason, we change the semantics of ``hashSubtreeCommitment`` in leaf nodes to commit
+
+Personalized BLAKE2b-256 was selected to match existing ZCash conventions. Adding the consensus branch ID to the hash personalization string ensures that valid nodes from one branch cannot be used to make false statements about parallel consensus branches.
+
+FlyClient Requirements and Recommendations
+===============================================
+These commitments enable FlyClient in the variable-difficulty model. Specifically, they allow light clients to reason about application of the difficulty adjustment algorithm over a range of blocks. They were chosen via discussion with an author of the FlyClient paper.
+
+- ``nEarliestTimestamp``
+- ``nLatestTimestamp``
+- ``nEarliestTarget``
+- ``nLatestTarget``
+- ``nEarliestHeight``
+- ``nLatestHeight``
+- ``nSubTreeTotalWork``
+
+Non-FlyClient Commitments
+==========================
+
+Additional metadata commitments were chosen primarily to improve light client security guarantees. We specified commitments where we could see an obvious security benefit, but there may be other useful metadata that we missed. We're interested in feedback and suggestions from the implementers of the current light client.
+
+We considered adding a commitment to the nullifier vector at each block. We would appreciate comments from light client teams on the utility of this commitment, as well as the proper serialization and commitment format for the nullifier vector.
+
+- ``hashEarliestSaplingRoot``
+
+  * Committing to the earliest Sapling root of a range of blocks allows light clients to check the consistency of treestate transitions over a range of blocks, without recalculating the root from genesis.
+
+- ``hashLatestSaplingRoot``
+
+  * This commitment serves the same purpose as ``hashFinalSaplingRoot`` in current Sapling semantics.
+  * However, because the MMR tree commits to blocks ``B_x ... B_(n-1)``, the latest commitment will descrie the final treestate of the previous block, rather than the current block.
```

Comment:
```suggestion
  * However, because the MMR tree commits to blocks ``B_x ... B_(n-1)``, the latest commitment will describe the final treestate of the previous block, rather than the current block.
```